go/pkg/pass1/print-group.1.md
# print-group 1 "" Netspoc "User Manual"

# NAME

print-group - Show elements of a netspoc group definition

# SYNOPSIS

print-group [options] FILE|DIR "group:name,..."

# DESCRIPTION

This program prints the elements of a Netspoc group.
It reads a group name from command line
and by default, shows a line with IP/prefixlen and name for each element.

Group is a named group, some automatic group, some named object or
a union or intersection or complement of simpler objects.

IP address, name and additional columns are separated by TAB character.

Name and IP address of a dual stack object is shown as two lines:

- first line with IPv4 address and name,
- second line with IPv6 address and same name.

# OPTIONS

**-n**, **--name**
:   Show only name of elements.

**-i**, **--ip**
:   Show only IP address of elements.

**-o**, **--owner**
:   Show owner of elements in additional column.

**-a**, **-admins**
:   Show admins of elements as comma separated list in additional column.

**-u**, **--unused**
:   Show only elements not used in any rules.

**--nat** name
:   Uses network:name as reference when resolving IP address in a NAT environment.

**-q**, **--quiet**
:   Don't print progress messages.

**-h**, **--help**
:   Print a brief help message and exit.

# EXAMPLES

Show names of elements of `group:g`
of Netspoc configuration in directory `netspoc/`:

`print-group --name netspoc/ group:g`

Show names and owners of elements of `group:a` without elements of `group:b`:

`print-group --name --owner netspoc/ 'group:a &! group:b'`

Show IP addresses of all hosts inside `area:x`:

`print-group --ip netspoc/ 'host:[area:x]'`

Show IP address and name of IPv6 networks inside `area:x`:

`print-group netspoc/ 'network:[any:[ip6=::/0 & area:x]]'`

# COPYRIGHT AND DISCLAIMER

(c) 2025 by Heinz Knutzen, heinz.knutzen@googlemail.com

This program is part of Netspoc, a Network Security Policy Compiler.
http://hknutzen.github.com/Netspoc

This program is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; either version 2 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY OR FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License along
with this program; if not, write to the Free Software Foundation, Inc.,
51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
*/


Changes
{{$NEXT}}

 - export-netvis: export Dualstack-Address of network
   in field Address2 of node.

6.079     2025-03-17 12:53:54+01:00 Europe/Berlin

 - If service has some owner with name "DA_*",
   this service is marked as visible for all owners with name "DA_*"
   in exported data for Netspoc-Web.
 - Command 'print-path' now accepts dual stack objects.

6.078     2025-02-24 14:42:54+01:00 Europe/Berlin

 - New commandline programs export-netvis and print-path have been added.

    These tools output information as JSON to be used in
    a graph visualization.

    export-netvis creates a list of all networks and routers
    of the given topology with information useful for visualizing
    the topology.

    print-path analyzes the given topology and creates a list of
    names of network nodes (routers and networks) between source and
    destination passed via command line.
    To create the list, print-path uses the same path-walk algorithm
    as used in netspoc main program to create ruleset for devices.

 - Aggregate with IPv4 IP 0.0.0.0/0 no longer matches IPv6 addresses.
   Aggregate with IPv6 IP ::/0 no longer matches IPv4 addresses.
 - Fixed split service from auto intf with IPv4, IPv6
 - Command 'export-netspoc' no longer strips comments and ";"
   from description in details of exported service.

6.077     2025-01-27 11:24:26+01:00 Europe/Berlin

 - Allow dual stack service to have protocols icmp and icmpv6 together.
   Generated rules automatically will use icmp for IPv4 and icmpv6 for IPv6.
 - Removed support for separate IPv6 topology in subdirectory 'ipv6/'.
 - Removed options '--ipv6', '-6'.
 - Removed support for NAT of IPv6 objects.
   After changing attributes of IPv6 addresses from 'ip' to 'ip6',
   supporting NAT for IPv6 would need to change NAT attributes
   to 'ip6', 'hidden6', 'identity6' as well.
   This seems to be superflous, since nobody uses NAT for IPv6.
 - Fixed duplicate hosts in print-group from wrong IPv4 network
   in IPv6 aggregate.
 - Fixed panic in print-group from wrong IPv4 network in IPv6 hosts
   generated by 'auto_ipv6_hosts'.
 - Fixed panic in export-netspoc with rule from IPv4 network to
   combined IPv4/v6 auto interface.
 - Changed 'format-netspoc' to print attributes "ip" and "ip6" in
   one line, if object has no more than 3 attributes.

6.076     2025-01-10 09:39:54+01:00 Europe/Berlin

 - Fixed dual stack bridge. IPv4 and IPv6 interfaces were accidently
   processed together.

6.075     2024-12-19 11:53:37+01:00 Europe/Berlin

 - Fix: Attribute 'policy_distribution_point' at dual stack router
   is only used for matching IPv4/IPv6 part of this router.
 - Fixed command 'export-netspoc'.
   Show IPv4 areas for combined zone. Must not accidently overwrite
   IPv4 areas by IPv6 areas, when combined zones have same name.

6.074     2024-12-16 11:52:23+01:00 Europe/Berlin

 - Fixed wrong error on duplicate interface from router in ipv6/ directory.

6.073     2024-12-16 10:16:33+01:00 Europe/Berlin

 - Added support for dual stack objects, having both, IPv4 and IPv6 addresses.
   This simplifies the modeling of a dual stack toplology.
   It is no longer neccessary to model a separate IPv4 and IPv6 topology.
   Rules between dual stack objects will generate ACls for IPv4 and IPv6.
   Rules between dual stack object and pure IPv4 object will silently ignore
   IPv6 address and generate only ACL for IPv4.
   New attributes have been introduced to define dual stack objects:
   - 'ip6' at network, host, interface and aggregate.
   - 'range6' at host.
   - 'unnumbered6' at network and interface.
   - 'negotiated6' at interface.
   Other changes resulting from use of dual stack objects:
   - If dual stack objects are used as border or inclusive border of an area,
     this defines two areas with identical name:
     one in IPv4 topology and one in IPv6 topology.
   - If dual stack objects are used as interfaces of an pathrestriction,
     this also defines two pathrestrictions in IPv4 and IPv6 topology.
     If the second pathrestriction has only one interface or only interfaces
     outside of a loop, it is silently ignored.
   - New attributes 'ipv4_only' and 'ipv6_only' may be used at service or area.
     This will enable only IPv4 or IPv6 part of dual stack objects.
   - The following attributes are applied only to IPv4 part
     if used in dual stack objects:
     'nat', 'bind_nat', 'subnet_of, 'hub', 'spoke'.
   - Command 'export-netspoc' has been changed to show IPv6 address of
     dual stack objects in attribute 'ip6'.
 - Added new attribute 'auto_ipv6_hosts' to automatically generate
   dual stack hosts from pure IPv4 hosts.
   It will generate IPv6 addresses for hosts
   by combining its IPv4 adress with the IPv6 address of its network.
   These attribute values are provided:
   - auto_ipv6_hosts = readable
     network: ip6 = 2001:db8:1:1::/64;
     hosts: ip = 172.17.1.48;
     ==> 2001:db8:1:1:172:17:1:48
   - auto_ipv6_hosts = binary
     network: ip6 = 2001:db8:1:1::/64;
     hosts: ip = 172.17.1.48;
     ==> 2001:db8:1:1::ac11:130
   - auto_ipv6_hosts = none
     No IPv6 address is generated
   This attribute can be used at network, area and host.
 - Fixed command 'cut-netspoc':
   Full path to management_instance is marked now.

6.072     2024-09-05 10:22:13+02:00 Europe/Berlin

 - Fixed command 'remove-from-netspoc'.
   A service is no longer removed accidently, if only some rules of
   a service become empty, but other rules remain.
 - Message "swap objects of 'user' and objects of rules" is shown
   even if no attribute multi_owner is given.
 - Fixed check for NAT network being subnet of some other network.
   If multiple networks have dynamic NAT to same address,
   now a warning is given for each network.
 - Fixed check for useless attribute 'subnet_of'
   where supernet has attribute 'has_subnets'.

6.071     2024-07-29 16:47:58+02:00 Europe/Berlin

 - Fixed generated ACL at router R with managed=local for rule where

go/pkg/pass1/print-group.go
package pass1

import (
	"fmt"
	"io"
	"net/netip"
	"strings"

	"github.com/hknutzen/Netspoc/go/pkg/conf"
	"github.com/hknutzen/Netspoc/go/pkg/oslink"
	"github.com/hknutzen/Netspoc/go/pkg/parser"
	"github.com/spf13/pflag"
)

// Print IP address of obj in context of natMap.
func printAddress(obj groupObj, nm natMap) string {
	netAddr := func(n *network) string {
		return prefixCode(n.ipp)
	}
	dynamicAddr := func(m map[string]netip.Addr, n *network) string {
		tag := n.natTag
		if ip, found := m[tag]; found {

			// Single static NAT IP for this object.
			return ip.String()
		} else if n.hidden {
			return "hidden"
		} else {

			// Dynamic NAT, take whole network.
			return netAddr(n)
		}
	}

	// Take higher bits from network NAT, lower bits from original IP.
	// This works with and without NAT.
	natAddr := func(ip netip.Addr, n *network) string {
		return mergeIP(ip, n).String()
	}

	switch x := obj.(type) {
	case *network:
		n := getNatNetwork(x, nm)
		if n.ipType == unnumberedIP {
			return "unnumbered"
		}
		if n.hidden {
			return "hidden"
		}
		return netAddr(n)
	case *host:
		n := getNatNetwork(x.network, nm)
		if n.dynamic {
			return dynamicAddr(x.nat, n)
		}
		if ip := x.ip; ip.IsValid() {
			return natAddr(ip, n)
		}
		return natAddr(x.ipRange.From(), n) + "-" + natAddr(x.ipRange.To(), n)
	case *routerIntf:
		n := getNatNetwork(x.network, nm)
		if n.dynamic {
			return dynamicAddr(x.nat, n)
		}
		switch x.ipType {
		case unnumberedIP:
			return "unnumbered"
		case shortIP:
			return "short"
		case bridgedIP:
			return "bridged"
		case negotiatedIP:
			// Take whole network.
			return netAddr(n)
		}
		return natAddr(x.ip, n)
	case *autoIntf:
		return "unknown"
	}
	return ""
}

func (c *spoc) printGroup(
	stdout io.Writer,
	path, group, natNet string,
	showIP, showName, showOwner, showAdmins, showUnused bool) {

	if !(showIP || showName) {
		showIP = true
		showName = true
	}
	parsed, err := parser.ParseUnion([]byte(group))
	if err != nil {
		c.abort("%v", err)
	}
	c.readNetspoc(path)
	c.setZone()
	c.stopOnErr()

	// Find network for resolving NAT addresses.
	var natMap natMap
	if natNet != "" {
		c.distributeNatInfo()
		c.stopOnErr()
		natNet = strings.TrimPrefix(natNet, "network:")
		if net := c.symTable.network[natNet]; net != nil {
			natMap = net.zone.natDomain.natMap
		} else {
			c.abort("Unknown network:%s of option '--nat'", natNet)
		}
	}

	// Prepare finding unused objects by marking used objects.
	used := make(map[groupObj]bool)
	if showUnused {
		c.setPath()
		sRules := c.normalizeServices()
		c.stopOnErr()
		process := func(rules []*serviceRule) {
			for _, rule := range rules {
				processObjects := func(group []srvObj) {
					for _, obj := range group {
						if gOb, ok := obj.(groupObj); ok {
							used[gOb] = true
						}
						switch x := obj.(type) {
						case *host:
							used[x.network] = true
						case *routerIntf:
							used[x.network] = true
						case *network:
							if x.hasIdHosts {
								for _, h := range x.hosts {
									used[h] = true
								}
							}
						}
					}
				}
				processObjects(rule.src)
				processObjects(rule.dst)
			}
		}
		process(sRules.permit)
		process(sRules.deny)
	}

	// Expand group definition.
	elements := c.expandGroup(parsed, "print-group", true)

	if showUnused {
		j := 0
		for _, ob := range elements {
			if !used[ob] {
				elements[j] = ob
				j++
			}
		}
		elements = elements[:j]
	}

	// Print IP address, name, owner, admins.
	//
	// Duplicate lines can result from
	// - combined IPv4/IPv6 objects and
	// - duplicated zones in zone cluster.
	seen := make(map[string]bool)
	for _, ob := range elements {
		var result stringList
		if showIP {
			result.push(printAddress(ob, natMap))
		}
		if showName {
			result.push(ob.String())
		}
		if showOwner || showAdmins {
			var ow *owner
			switch x := ob.(type) {
			case srvObj:
				ow = x.getOwner()
			case *area:
				ow = x.owner
			}
			oName := "none"
			admins := ""
			if ow != nil {
				oName = ow.name
				admins = strings.Join(ow.admins, ",")
			}
			if showOwner {
				result.push(oName)
			}
			if showAdmins {
				result.push(admins)
			}
		}
		line := strings.Join(result, "\t")
		if !seen[line] {
			fmt.Fprintln(stdout, line)
			seen[line] = true
		}
	}
}

func PrintGroupMain(d oslink.Data) int {
	fs := pflag.NewFlagSet(d.Args[0], pflag.ContinueOnError)

	// Setup custom usage function.
	fs.Usage = func() {
		fmt.Fprintf(d.Stderr,
			"Usage: %s [options] FILE|DIR 'group:name,...'\n%s",
			d.Args[0], fs.FlagUsages())
	}

	// Command line flags
	quiet := fs.BoolP("quiet", "q", false, "Don't print progress messages")

	nat := fs.String("nat", "",
		"Use network:name as reference when resolving IP address")
	unused := fs.BoolP("unused", "u", false,
		"Show only elements not used in any rules")
	name := fs.BoolP("name", "n", false, "Show only name of elements")
	ip := fs.BoolP("ip", "i", false, "Show only IP address of elements")
	owner := fs.BoolP("owner", "o", false, "Show owner of elements")
	admins := fs.BoolP("admins", "a", false,
		"Show admins of elements as comma separated list")
	if err := fs.Parse(d.Args[1:]); err != nil {
		if err == pflag.ErrHelp {
			return 1
		}
		fmt.Fprintf(d.Stderr, "Error: %s\n", err)
		fs.Usage()
		return 1
	}

	// Argument processing
	args := fs.Args()
	if len(args) != 2 {
		fs.Usage()
		return 1
	}
	path := args[0]
	group := args[1]

	cnf := conf.ConfigFromFile(path)
	cnf.Quiet = *quiet
	return toplevelSpoc(d, cnf, func(c *spoc) {
		c.printGroup(
			d.Stdout, path, group, *nat, *ip, *name, *owner, *admins, *unused)
	})
}
